File: library/src/test/java/io/github/anderscheow/validator/rules/common/RegexRuleTest.java
package io.github.anderscheow.validator.rules.common;

import android.support.annotation.StringRes;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import java.util.Locale;

import static org.junit.Assert.*;

public class RegexRuleTest {

    private static final String REGEX = "\\d+";

    private RegexRule regexRule;

    @Before
    public void setUp() throws Exception {
    }

    @After
    public void tearDown() throws Exception {
        regexRule = null;
    }

    @Test(expected = NullPointerException.class)
    public void validate_EmptySample_ThrowNullPointerException() throws Exception {
        regexRule = new RegexRule(REGEX) {};

        regexRule.validate(null);
    }

    @Test(expected = ClassCastException.class)
    public void validate_NotStringSample_ThrowClassCastException() throws Exception {
        regexRule = new RegexRule(REGEX) {};

        // Any value other than String
        regexRule.validate(123);
    }

    @Test
    public void validate_ValidSample_ReturnTrue() throws Exception {
        regexRule = new RegexRule(REGEX) {};

        String sample = "123";

        assertTrue(regexRule.validate(sample));
    }

    @Test
    public void validate_InvalidSample_ReturnFalse() throws Exception {
        regexRule = new RegexRule(REGEX) {};

        String sample = "TEST";

        assertFalse(regexRule.validate(sample));
    }

    @Test
    public void errorMessage_DefaultErrorMessage() throws Exception {
        String errorMessage = "Does not match regex rule";

        regexRule = new RegexRule(REGEX) {};

        assertEquals(errorMessage, regexRule.errorMessage());
    }

    @Test
    public void errorMessage_CustomErrorMessage() throws Exception {
        String errorMessage = "This is a custom error message";

        regexRule = new RegexRule(REGEX, errorMessage) {};

        assertEquals(errorMessage, regexRule.errorMessage());
    }

    @Test
    public void errorMessage_CustomErrorRes() throws Exception {
        @StringRes int errorRes = 0;

        regexRule = new RegexRule(REGEX, errorRes) {};

        assertEquals(errorRes, regexRule.errorRes());
    }
}

File: library/src/test/java/io/github/anderscheow/validator/rules/common/RegexRuleTest.kt
package io.github.anderscheow.validator.rules.common

import androidx.annotation.StringRes
import org.junit.After
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test

class RegexRuleTest {

    private lateinit var regexRule: RegexRule

    @Before
    @Throws(Exception::class)
    fun setUp() {
    }

    @After
    @Throws(Exception::class)
    fun tearDown() {
    }

    @Test(expected = NullPointerException::class)
    @Throws(Exception::class)
    fun validate_EmptySample_ThrowNullPointerException() {
        regexRule = object : RegexRule(REGEX) {
        }

        regexRule.validate(null)
    }

    @Test
    @Throws(Exception::class)
    fun validate_ValidSample_ReturnTrue() {
        regexRule = object : RegexRule(REGEX) {
        }

        val sample = "123"

        assertTrue(regexRule.validate(sample))
    }

    @Test
    @Throws(Exception::class)
    fun validate_InvalidSample_ReturnFalse() {
        regexRule = object : RegexRule(REGEX) {
        }

        val sample = "TEST"

        assertFalse(regexRule.validate(sample))
    }

    @Test
    @Throws(Exception::class)
    fun errorMessage_DefaultErrorMessage() {
        val errorMessage = "Does not match regex rule"

        regexRule = object : RegexRule(REGEX) {
        }

        assertEquals(errorMessage, regexRule.getErrorMessage())
    }

    @Test
    @Throws(Exception::class)
    fun errorMessage_CustomErrorMessage() {
        val errorMessage = "This is a custom error message"

        regexRule = object : RegexRule(REGEX, errorMessage) {
        }

        assertEquals(errorMessage, regexRule.getErrorMessage())
    }

    @Test
    @Throws(Exception::class)
    fun errorMessage_CustomErrorRes() {
        @StringRes val errorRes = 0

        regexRule = object : RegexRule(REGEX, errorRes) {
        }

        assertEquals(errorRes, regexRule.getErrorRes())
    }

    companion object {

        private const val REGEX = "\\d+"
    }
}